Implement per-site index (fixes #1801)

Mohammed90 opened the pull request:
### 1. What does this change do, exactly?
Adds per-site index capability while maintaining the a list of possible index file names

### 2. Please link to the relevant issues.
Fixes #1801 

### 3. Which documentation changes (if any) need to be made because of this PR?
The index plugin docs. Currently it's not clear whether it's per-site or not, which is what the linked issue is about.

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mholt/caddy?pullRequest=1906) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/mholt/caddy?pullRequest=1906) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mholt/caddy?pullRequest=1906) it.</sub>

francislavoie commented on caddyhttp/index/index.go:
```diff
@@ -29,6 +29,8 @@ func init() {
 func setupIndex(c *caddy.Controller) error {
 	var index []string
 
+	cnfg := httpserver.GetConfig(c)
```
I think the usual naming is `cfg` for these

Mohammed90 replied:
Done

mholt commented:
Awesome! I'm gonna look at this soon after my headache goes away! 🕺

Mohammed90 commented:
@elcore, done, though I feel like they're tautological tests. I think the best way to effectively test this is with integration tests.

mholt commented:
These tests aren't that unwieldy -- I wouldn't say it's _necessary_ to use table tests, but if you _want_ to in this case, you may. I'll do more of a review soon.

elcore commented:
> These tests aren't that unwieldy -- I wouldn't say it's necessary to use table tests, but if you want to in this case, you may. I'll do more of a review soon.

@mholt It is not necessary to use table tests, but it simplifies the code etc. etc.

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
+	// should receive staticfiles.DefaultIndexPages slice.
+	if len(siteConfig.IndexPages) != len(siteConfig.IndexPages) {
+		t.Errorf("Expected 3 values, got %v", len(siteConfig.IndexPages))
+	}
```
```
// In case the index directive is not used, the virtual host
// should receive staticfiles.DefaultIndexPages slice.
if len(siteConfig.IndexPages) != len(staticfiles.DefaultIndexPages) {
	t.Errorf("Expected %v values, got %v", len(staticfiles.DefaultIndexPages), len(siteConfig.IndexPages))
}
```

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -17,8 +17,10 @@ package index
+
+	"github.com/mholt/caddy"
+	"github.com/mholt/caddy/caddyhttp/httpserver"
 )
```
Delete whitespace and gofmt to rearrange imports

```
import (
        "testing"

        "github.com/mholt/caddy"
        "github.com/mholt/caddy/caddyhttp/httpserver"
        "github.com/mholt/caddy/caddyhttp/staticfiles"
)

```

elcore commented on caddyhttp/httpserver/plugin.go:
```diff
@@ -25,6 +25,8 @@ import (
 	"time"
 
+	"github.com/mholt/caddy/caddyhttp/staticfiles"
+
```
Delete whitespace and gofmt to rearrange imports

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
 }
+
+func TestMultiSiteIndexWithEitherHasDefault(t *testing.T) {
+
```
Delete whitespace

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
+}
+
+func TestMultiSiteIndexWithBothCustomized(t *testing.T) {
+
```
Delete whitespace

elcore commented on caddyhttp/httpserver/plugin.go:
```diff
@@ -25,6 +25,8 @@ import (
 	"time"
 
+	"github.com/mholt/caddy/caddyhttp/staticfiles"
+
```
Delete whitespace and gofmt to rearrange imports

Mohammed90 replied:
Done

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -17,8 +17,10 @@ package index
+
+	"github.com/mholt/caddy"
+	"github.com/mholt/caddy/caddyhttp/httpserver"
 )
```
Delete whitespace and gofmt to rearrange imports

```
import (
        "testing"

        "github.com/mholt/caddy"
        "github.com/mholt/caddy/caddyhttp/httpserver"
        "github.com/mholt/caddy/caddyhttp/staticfiles"
)

```

Mohammed90 replied:
Done

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
 }
+
+func TestMultiSiteIndexWithEitherHasDefault(t *testing.T) {
+
```
Delete whitespace

Mohammed90 replied:
Done

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
+	// should receive staticfiles.DefaultIndexPages slice.
+	if len(siteConfig.IndexPages) != len(siteConfig.IndexPages) {
+		t.Errorf("Expected 3 values, got %v", len(siteConfig.IndexPages))
+	}
```
The code does not reflect the comment

```
// In case the index directive is not used, the virtual host
// should receive staticfiles.DefaultIndexPages slice.
if len(siteConfig.IndexPages) != len(staticfiles.DefaultIndexPages) {
	t.Errorf("Expected %d values, got %d", len(staticfiles.DefaultIndexPages), len(siteConfig.IndexPages))
}
```

Mohammed90 replied:
I added more to the comment to clarify my intention

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +42,85 @@ func TestIndex(t *testing.T) {
+}
+
+func TestMultiSiteIndexWithBothCustomized(t *testing.T) {
+
```
Delete whitespace

Mohammed90 replied:
Done

Mohammed90 commented:
@elcore, I see where the misunderstanding is. Those 10 lines are not two test cases. The test is to verify the isolation of index page configuration across different sites. I renamed the test and added comments to make it clearer.

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +41,85 @@ func TestIndex(t *testing.T) {
+	// and the values, as checked in the upcoming loop, should match.
+	if len(siteConfig.IndexPages) != len(siteConfig.IndexPages) {
+		t.Errorf("Expected 3 values, got %v", len(siteConfig.IndexPages))
+	}
```
`len(siteConfig.IndexPages) != len(siteConfig.IndexPages)` : This is a paradox

`Expected 3 values` : This is wrong as currently 6 values are expected

Reference: https://play.golang.org/p/aBIYMl2h--

> In case the index directive is not used, the virtual host should receive staticfiles.DefaultIndexPage. The length, as checked here, [...] should match.

As far as I understand your comment, the code should look like this:

```
if len(siteConfig.IndexPages) != len(staticfiles.DefaultIndexPages) {
	t.Errorf("Expected %d values, got %d", len(staticfiles.DefaultIndexPages), len(siteConfig.IndexPages))
}
```

ping @Mohammed90 , @mholt

elcore commented on caddyhttp/index/index_test.go:
```diff
@@ -40,14 +41,85 @@ func TestIndex(t *testing.T) {
+	// and the values, as checked in the upcoming loop, should match.
+	if len(siteConfig.IndexPages) != len(siteConfig.IndexPages) {
+		t.Errorf("Expected 3 values, got %v", len(siteConfig.IndexPages))
+	}
```
`Expected 3 values` : This is wrong as currently 6 values are expected

Reference: https://play.golang.org/p/aBIYMl2h--

> In case the index directive is not used, the virtual host should receive staticfiles.DefaultIndexPage. The length, as checked here, [...] should match.

As far as I understand your comment, the code should look like this:

```
if len(siteConfig.IndexPages) != len(staticfiles.DefaultIndexPages) {
	t.Errorf("Expected %d values, got %d", len(staticfiles.DefaultIndexPages), len(siteConfig.IndexPages))
}
```

ping @Mohammed90 , @mholt

Mohammed90 replied:
Oops! Yeah, this one is embarrassing 😅 Fixed

elcore replied:
:smile:

elcore commented:
@Mohammed90 You've done a great job so far, I appreciate your contribution!!

Mohammed90 commented:
@elcore do you still think I need to add more tests? If so, can you point me at what conditions am I missing in the current tests.

elcore commented:
>@elcore do you still think I need to add more tests? If so, can you point me at what conditions am I missing in the current tests.

@Mohammed90 I do not like the way you test, but the code is fine
